File: app/src/main/java/ua/hospes/rtm/ui/teams/TeamsPresenter.java
package ua.hospes.rtm.ui.teams;

import javax.inject.Inject;

import ua.hospes.rtm.core.ui.BasePresenter;
import ua.hospes.rtm.domain.team.TeamsInteractor;
import ua.hospes.rtm.utils.RxUtils;

/**
 * @author Andrew Khloponin
 */
class TeamsPresenter extends BasePresenter<TeamsContract.View> {
    private final TeamsInteractor interactor;

    @Inject
    TeamsPresenter(TeamsInteractor interactor) {
        this.interactor = interactor;
    }

    @Override
    public void attachView(TeamsContract.View view) {
        super.attachView(view);

        RxUtils.manage(this, interactor.listen()
                .compose(RxUtils.applySchedulers())
                .subscribe(list -> getView().updateTeams(list), Throwable::printStackTrace));
    }

    @Override
    public void detachView() {
        super.detachView();
        RxUtils.unsubscribe(this);
    }


    public void removeAll() {
        RxUtils.manage(this, interactor.removeAll()
                .compose(RxUtils.applySchedulersSingle())
                .subscribe(result -> {}, Throwable::printStackTrace));
    }
}

File: app/src/main/java/ua/hospes/rtm/ui/teams/TeamsPresenter.kt
package ua.hospes.rtm.ui.teams

import androidx.lifecycle.Lifecycle
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import ua.hospes.rtm.core.Presenter
import ua.hospes.rtm.data.TeamsRepository
import javax.inject.Inject

class TeamsPresenter @Inject constructor(private val repo: TeamsRepository) : Presenter<TeamsContract.View>() {

    override fun attachView(view: TeamsContract.View?, lc: Lifecycle) {
        super.attachView(view, lc)

        launch(Dispatchers.Main) { repo.listen().collect { view?.onData(it) } }
    }

    override fun onError(throwable: Throwable) = view?.onError(throwable) ?: Unit
    override fun onUnexpectedError(throwable: Throwable) = view?.onError(throwable) ?: Unit


    fun removeAll() = launch { repo.clear() }
}